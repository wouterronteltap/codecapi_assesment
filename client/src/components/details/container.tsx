import React, { useEffect } from 'react';
import { History } from "history";

import { useAppState } from "../../context";
import { UserDetails } from './details';
import { useApi } from "../../hooks/useApi";

import { CONSTANTS } from "../../utils/contants";
import {Typography} from "@material-ui/core";

interface IProps {
    login: string,
    history: History
}

const DetailsContainer = (props: IProps) => {
    const { login, history } = props;
    const { data, isError, isLoading, error } = useAppState();
    const url = CONSTANTS.GET_USER_DETAILS_URL+login;
    const [setUrl] = useApi(url);
    useEffect(() => {
        setUrl(url)
    }, [login]);
    return(
        <>
            {data && !isLoading && !isError && <UserDetails data={ data } history={history}/>}
            {isLoading && <Typography variant={'body1'}>loading ...</Typography>}
            {isError && <Typography variant={'body1'}>{error}</Typography>}
        </>
    )
};

export { DetailsContainer };