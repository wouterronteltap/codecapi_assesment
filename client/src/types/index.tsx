import {Dispatch} from "react";

// TODO REWRITE PER ACTION
export interface IAction {
    type: string;
    // TODO => PROPER TYPE FOR PAYLOAD
    payload?: any;
}

export interface ILinks {
    last?: string,
    next?: string
}

export type IUser = {
    login: string,
    id: number,
    score?: number,
    avatar_url?: string,
    company?: string,
    email?: string,
    followers: number,
    following: number,
    name?: string,
    location?: string,
    bio?: string,
    blog?: string,
    repos_url: string,
    html_url: string

}

export interface IUsersResponse {
    users:  Array<IUser>,
    links: ILinks,
    total: number
}

export interface IState<T> {
    query: string,
    currentPage: number,
    total: number,
    lastPage: number,
    isLoading: boolean;
    isError: boolean;
    data?: T;
}

export interface IDispatch {
    dispatch: Dispatch<IAction>;
}
