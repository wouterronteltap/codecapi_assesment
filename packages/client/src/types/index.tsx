import { Dispatch } from "react";
import { AxiosResponse, AxiosError } from "axios";

type Actions =
  | "FETCH_INIT"
  | "FETCH_SUCCESS"
  | "FETCH_FAILURE"
  | "SET_PAGE"
  | "SET_DEBOUNCED_QUERY";

export interface IAction {
  type: Actions;
  payload: any;
}

export interface ILinks {
  last?: string;
  next?: string;
}

export type IUser = {
  login: string;
  id: number;
  score?: number;
  avatar_url?: string;
  company?: string;
  email?: string;
  followers: number;
  following: number;
  name?: string;
  location?: string;
  bio?: string;
  repos_url: string;
  html_url: string;
};

export interface IResponse extends AxiosResponse {
  users: Array<IUser>;
  links?: ILinks;
  total?: number;
}

export interface IState {
  query: string;
  currentPage: number;
  total: number;
  lastPage: number;
  isLoading: boolean;
  isError: boolean;
  data: IResponse | null;
  error: AxiosError | null;
}

export interface IDispatch {
  dispatch: Dispatch<IAction>;
}
