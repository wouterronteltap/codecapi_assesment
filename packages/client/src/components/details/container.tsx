import React, { useEffect } from 'react'

import { useAppState } from '../../context'
import { UserDetails } from '../../components/details/details'
import { useApi } from '../../hooks/useApi'

import { CONSTANTS } from '../../utils/contants'
import { Typography } from '@material-ui/core'

interface IProps {
  login: string
}

const DetailsContainer = (props: IProps) => {
  const { login } = props
  const { data, isError, isLoading, error } = useAppState()
  const url = CONSTANTS.GET_USER_DETAILS_URL(login)
  const [setUrl] = useApi(url)
  useEffect(() => {
    setUrl(url)
  }, [login])
  return (
    <>
      {data && !isLoading && !isError && <UserDetails data={data} />}
      {isLoading && <Typography variant={'body1'}>loading ...</Typography>}
      {isError && <Typography variant={'body1'}>{error}</Typography>}
    </>
  )
}

export { DetailsContainer }
