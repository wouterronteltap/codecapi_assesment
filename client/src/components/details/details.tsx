import React from 'react';
import { Avatar, List, ListItem, ListItemText, ListItemAvatar, Button } from '@material-ui/core';
import { History } from "history";
import { IResponse } from "../../types";

interface IProps {
    data: IResponse,
    history: History
}

const UserDetails = React.memo((props: IProps) => {
    const { users }  = props.data;
    const {login, avatar_url, company, followers,following, name, location, bio, repos_url, html_url } = users[0];
    return(
        <List dense={true}>
            <ListItem>
                <ListItemAvatar>
                    <Avatar src={avatar_url} />
                </ListItemAvatar>
            </ListItem>
            <ListItem>
                <ListItemText primary={"Username: "} secondary={login}/>
            </ListItem>
            <ListItem>
                <ListItemText primary={"Name: "} secondary={name}/>
            </ListItem>
            {location &&
                <ListItem>
                    <ListItemText primary={"Location: "} secondary={name}/>
                </ListItem>}
            <ListItem>
                <ListItemText primary={"Profile: "} secondary={name}/>
            </ListItem>
            <ListItem>
                <a href={html_url}>
                    <ListItemText primary={"Github url: "} secondary={html_url}/>
                </a>
            </ListItem>
            <ListItem>
                <a href={repos_url}>
                    <ListItemText primary={"Repos url: "} secondary={repos_url}/>
                </a>
            </ListItem>
            {company &&
                <ListItem>
                    <ListItemText primary={"Company: "} secondary={company}/>
                </ListItem>}
            {bio &&
                <ListItem>
                    <ListItemText primary={"Bio: "} secondary={bio}/>
                </ListItem>}
            <ListItem>
                <ListItemText primary={"Followers: "} secondary={followers}/>
            </ListItem>
            <ListItem>
                <ListItemText primary={"Following: "} secondary={following}/>
            </ListItem>
            <Button variant={'outlined'} onClick={() => props.history.goBack()}>
                back
            </Button>
        </List>
    )
});

export { UserDetails };