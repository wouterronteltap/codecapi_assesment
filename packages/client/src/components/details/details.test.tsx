import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { UserDetails, IProps } from "./details";
import { IResponse } from "../../types";

const data: IResponse = {
  users: [
    {
      id: 1,
      name: "Wouter",
      score: 9,
      login: "Woutertje",
      avatar_url: "fakepath.png",
      followers: 19,
      following: 10,
      repos_url: "/fakepath/repos",
      html_url: "/fakepath"
    }
  ],
  data: {},
  status: 0,
  headers: {},
  config: {},
  statusText: ""
};

const userMock: IProps = {
  data: data
};

describe("The CustomListItem Component", () => {
  afterEach(cleanup);
  test("renders correctly", () => {
    const { asFragment } = render(<UserDetails data={data} />, {
      wrapper: MemoryRouter
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
