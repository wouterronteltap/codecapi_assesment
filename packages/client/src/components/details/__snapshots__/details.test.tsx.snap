// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The CustomListItem Component renders correctly 1`] = `
<DocumentFragment>
  <h6
    class="MuiTypography-root MuiTypography-h6"
  >
    User Details:
  </h6>
  <ul
    class="MuiList-root MuiList-dense MuiList-padding"
  >
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
    >
      <div
        class="MuiListItemAvatar-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle"
        >
          <img
            class="MuiAvatar-img"
            src="fakepath.png"
          />
        </div>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
    >
      <div
        class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body2"
        >
          Username: 
        </span>
        <p
          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          Woutertje
        </p>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
    >
      <div
        class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body2"
        >
          Name: 
        </span>
        <p
          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          Wouter
        </p>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
    >
      <div
        class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body2"
        >
          Profile: 
        </span>
        <p
          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          Wouter
        </p>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
    >
      <a
        href="/fakepath"
      >
        <div
          class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body2"
          >
            Github url: 
          </span>
          <p
            class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            /fakepath
          </p>
        </div>
      </a>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
    >
      <a
        href="/fakepath/repos"
      >
        <div
          class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body2"
          >
            Repos url: 
          </span>
          <p
            class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            /fakepath/repos
          </p>
        </div>
      </a>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
    >
      <div
        class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body2"
        >
          Followers: 
        </span>
        <p
          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          19
        </p>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
    >
      <div
        class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body2"
        >
          Following: 
        </span>
        <p
          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          10
        </p>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
    >
      <a
        href="/"
      >
        back
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
